rustler_bigint/src/big_int.rs

```diff
@@ -0,0 +1,396 @@
+use rustler::types::atom::error;
+use rustler::{Decoder, Encoder, Env, Error, NifResult, Term};
+
+use num_bigint::Sign;
+
+/// Wrapper around num-bigint that implements [Decoder](rustler::Decoder) and [Encoder](rustler::Encoder) traits
+///
+/// ```rust
+/// use rustler_bigint::{BigInt, num_bigint};
+///
+/// // convert from and to the bigint wrapper
+/// let number = num_bigint::BigInt::from(12);
+/// let implements_decode_encode = BigInt::from(number.clone());
+/// let bigint = num_bigint::BigInt::from(implements_decode_encode);
+/// assert_eq!(number, bigint);
+/// ```
+///
+/// ## Examples
+///
+/// ```rust
+/// use rustler_bigint::BigInt;
+///
+/// #[rustler::nif]
+/// pub fn pow(base: BigInt, exponent: u32) -> BigInt {
+///   base.pow(exponent).into()
+/// }
+/// ```
+///
+/// ```rust
+/// use rustler::Binary;
+/// use rustler_bigint::{BigInt, num_bigint};
+///
+/// #[rustler::nif]
+/// pub fn binary_to_integer(binary: Binary) -> BigInt {
+///   num_bigint::BigInt::from_signed_bytes_be(binary.as_slice()).into()
+/// }
+/// ```
+///
+/// ```rust
+/// use rustler::{Binary, Env, NewBinary};
+/// use rustler_bigint::BigInt;
+///
+/// #[rustler::nif]
+/// pub fn integer_to_binary<'a>(env: Env<'a>, integer: BigInt) -> Binary<'a> {
+///   let bytes = integer.to_signed_bytes_be();
+///   let mut output = NewBinary::new(env, bytes.len());
+///   output.as_mut_slice().copy_from_slice(bytes.as_slice());
+///   output.into()
+/// }
+/// ```
+///
+#[derive(Debug, PartialEq, PartialOrd)]
+pub struct BigInt(num_bigint::BigInt);
+
+impl std::convert::From<num_bigint::BigInt> for BigInt {
+    fn from(big_int: num_bigint::BigInt) -> BigInt {
+        BigInt(big_int)
+    }
+}
+
+impl std::convert::From<BigInt> for num_bigint::BigInt {
+    fn from(big_int: BigInt) -> num_bigint::BigInt {
+        big_int.0
+    }
+}
+
+impl std::ops::Deref for BigInt {
+    type Target = num_bigint::BigInt;
+
+    fn deref(&self) -> &Self::Target {
+        &self.0
+    }
+}
+
+impl std::ops::DerefMut for BigInt {
+    fn deref_mut(&mut self) -> &mut Self::Target {
+        &mut self.0
+    }
+}
+
+fn decode_big_integer(input: &[u8]) -> NifResult<num_bigint::BigInt> {
```

Comment:
I think we should define constants for the magical numbers here:

```rust
// From https://www.erlang.org/doc/apps/erts/erl_ext_dist.html
const SMALL_INTEGER = 97u8;
const INTEGER = 98u8;
const  SMALL_BIG_EXT = 110u8;
const LARGE_BIG_EXT = 111u8;
```

Reply:
done